Use Awaited type to avoid Promise<Promise<T>>

urbanfly opened the pull request:
When debouncing a Promise-returning function, the returned function has a type of `Promise<Promise<T>>` which is not accurate - see https://github.com/microsoft/TypeScript/issues/27711

In TypeScript 4.5, a new `Awaited<T>` type was introduced to solve this problem.

- [docs](https://www.typescriptlang.org/docs/handbook/utility-types.html#awaitedtype)
- [release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#the-awaited-type-and-promise-improvements)

This PR simply wraps the `ReturnType<F>` with `Awaited<...>` in order to get an accurate return type for the created function.

Several packages were updated:

- Typescript bumped to 4.5 in order to access this type
- microbundle in order to build
- jest / ts-jest in order to test with this type
- tsd in order to validate the types

urbanfly committed:
Use Awaited type to avoide Promise<Promise<T>>

urbanfly committed:
add a test to debounce a Promise-returning function

chodorowicz commented:
Hey @urbanfly 
Sorry for the absurdly long delay with replying 🙇

The change makes sense, thanks. There's one issue though, the type for the callback functions `Options<Awaited<ReturnType<F>>>` doesn't match the existing behavior as the promises there are not flattened with `resolve` and result is passed directly to the callback function. I've added tests that document that behavior → https://github.com/chodorowicz/ts-debounce/pull/70/changes

if you're still around and interested in merging this change, I'm happy to do it after this issue is fixed.

urbanfly committed:
Remove the Awaited type from the `options` argument. Relevant tests are in #70

urbanfly commented:
I didn't originally get your point about the callback/options. I have removed `Awaited` from the `options` argument.

chodorowicz commented:
Thanks, looks good!

There's conflict with the base branch, which needs to get resolved first (both merge and rebase are fine)
Also, type test have issue 
```
  ✖  34:0  Parameter type () => Promise<number> is declared too wide for argument type DebouncedFunction<[], () => Promise<number>>.
 ```
`expectType` requires strict type equality. using `expectAssignable` from `tsd` should be good to solve the problem.

Happy to merge this after resolving this two issue.

urbanfly commented:
I've rebased and tried to run the tests. The Jest tests pass, but `test:types` fails on `tsd`:
> The type definition `dist/src/index.d.ts` does not exist at `/workspaces/ts-debounce/dist/src/index.d.ts`. Is the path correct? Create one and try again.

It seems like it depends on the `build` script to run first, but that also fails:
> <img width="981" height="209" alt="image" src="https://github.com/user-attachments/assets/6b3a958e-8365-46ec-aa02-6f9c2c3e78d6" />
> 
> Add [@babel/preset-flow](https://github.com/babel/babel/tree/main/packages/babel-preset-flow) to the `presets` section of your Babel config to enable transformation.
> If you want to leave it as-is, add [@babel/plugin-syntax-flow](https://github.com/babel/babel/tree/main/packages/babel-plugin-syntax-flow) to the 'plugins' section to enable parsing.
>
> If you already added the plugin for this syntax to your config, it's possible that your config isn't being loaded.
> You can re-run Babel with the `BABEL_SHOW_CONFIG_FOR` environment variable to show the loaded configuration:
> ```bash
> npx cross-env BABEL_SHOW_CONFIG_FOR=/workspaces/ts-debounce/src/index.ts <your build command>
> ```
> See https://babeljs.io/docs/configuration#print-effective-configs for more info.

I was able to work-around it locally by removing the `--noEmit` flag from `tsc` so that the `dist` folder gets populated (and then adjust the paths for `types` in `package.json` to match the expectation). So at this point, `npm run test` passes, but only with some uncommitted work-arounds.

chodorowicz commented:
The CI job, which includes `pnpm run test:types`, [is green](https://github.com/chodorowicz/ts-debounce/actions/runs/24084311104/job/70327593307?pr=68).

The last thing is to remove `package-lock.json` from the PR, as this project uses `pnpm` and `pnpm-lock.yaml`.

urbanfly committed:
Removed package-lock.json since this project uses pnpm and pnpm-lock.yaml

urbanfly committed:
Update @types/jest to version 29 in devDependencies

urbanfly commented on package.json:
```diff
@@ -29,13 +29,13 @@
-    "typescript": "^4.4.4"
+    "ts-jest": "^29",
+    "tsd": "^0.31",
+    "typescript": "^4.5"
```
@types/jest updated to match

urbanfly reviewed (commented):
I don't know how I "Started a review". I just replied to Copilots's comment re: `@types/jest` version within VS Code.

urbanfly committed:
Properly upgrade `@types/jest` using `pnpm install`

Updated Jest doesn't accept "modern" anymore. See https://jest-archive-august-2023.netlify.app/docs/28.x/upgrading-to-jest28/#jestusefaketimers

urbanfly committed:
declare jest types to remove red squiggles from IDE

urbanfly commented:
I saw the build failure and have now properly installed `@types/jest` to v29 so that the `pnpm-lock.json` file is updated correctly.

Updating to Jest v29 made `useFakeTimes("modern")` fail, so removed the parameter. See https://jest-archive-august-2023.netlify.app/docs/28.x/upgrading-to-jest28/#jestusefaketimers

chodorowicz commented:
@all-contributors add @urbanfly for code

chodorowicz commented:
Thanks for your contribution @urbanfly 🙇
